src/library.js

```diff
@@ -2845,6 +2845,47 @@ LibraryManager.library = {
   endgrent: function() { throw 'endgrent: TODO' },
   setgrent: function() { throw 'setgrent: TODO' },
 
+  // random.h
+
+  // TODO: consider allowing the API to get a parameter for the number of
+  // bytes.
+  $getRandomDevice: function() {
+    if (typeof crypto === 'object' && typeof crypto['getRandomValues'] === 'function') {
+      // for modern web browsers
+      var randomBuffer = new Uint8Array(1);
+      return function() { crypto.getRandomValues(randomBuffer); return randomBuffer[0]; };
+    } else
+#if ENVIRONMENT_MAY_BE_NODE
+    if (ENVIRONMENT_IS_NODE) {
+      // for nodejs with or without crypto support included
+      try {
+        var crypto_module = require('crypto');
+        // nodejs has crypto support
+        return function() { return crypto_module['randomBytes'](1)[0]; };
+      } catch (e) {
+        // nodejs doesn't have crypto support
+      }
+    }
+#endif // ENVIRONMENT_MAY_BE_NODE
+    // we couldn't find a proper implementation, as Math.random() is not suitable for /dev/random, see emscripten-core/emscripten/pull/7096
+#if ASSERTIONS
+    return function() { abort("no cryptographic support found for randomDevice. consider polyfilling it if you want to use something insecure like Math.random(), e.g. put this in a --pre-js: var crypto = { getRandomValues: function(array) { for (var i = 0; i < array.length; i++) array[i] = (Math.random()*256)|0 } };"); };
+#else
+    return function() { abort("randomDevice"); };
+#endif
+  },
+
+  getentropy__deps: ['$getRandomDevice'],
+  getentropy: function(buffer, size) {
+    if (!_getentropy.randomDevice) {
+      _getentropy.randomDevice = getRandomDevice();
+    }
```

Comment:
Would it makes sense to do this memorization in getRandomDevice itself?

Also if you made `_getentropy.randomDevice` into `$randomDevice` then you could maybe save a few bytes?

Reply:
I tried a few ways, but it's hard to reduce the code size below what it currently is.

Reply:
But we not memoize the getRandomDevice inner function?   And why not use global for the memoization?

Reply:
I see `getRandomDevice` as getting a random device. It returns a new function each time.

In theory each of those functions could have its own internal DRNG or such (see discussion in #10068). 

Callers can memoize when it makes sense. In the FS code we and in getentropy I think it does.

I do agree that if memoizing in `getRandomDevice` itself were smaller, it would be worth figuring out a proper name for that different API. I can't get it any smaller, though...